src/actions.rs

```diff
@@ -0,0 +1,412 @@
+                self.instructions.push(Instruction::Set(func.name));
+            }
+            GenericCoreAction::Delete(f, args) => {
+                let ResolvedCall::Func(func) = f else {
+                    panic!("Cannot delete primitive- should have been caught by typechecking!!!")
+                };
+                for arg in args {
+                    self.do_atom_term(arg);
+                }
+                self.instructions.push(Instruction::DeleteRow(func.name));
+            }
+            GenericCoreAction::Union(arg1, arg2) => {
+                self.do_atom_term(arg1);
+                self.do_atom_term(arg2);
+                self.instructions.push(Instruction::Union(2));
+            }
+            GenericCoreAction::Panic(msg) => {
+                self.instructions.push(Instruction::Panic(msg.clone()));
+            }
+        }
+    }
+
+    fn egraph(&self) -> &'a EGraph {
+        self.egraph
+    }
+
+    fn do_call(&mut self, f: &ResolvedCall, args: &[ResolvedAtomTerm]) {
+        for arg in args {
+            self.do_atom_term(arg);
+        }
+        match f {
+            ResolvedCall::Func(f) => self.do_function(f),
+            ResolvedCall::Primitive(p) => self.do_prim(p),
+        }
+    }
+
+    fn do_atom_term(&mut self, at: &ResolvedAtomTerm) {
+        match at {
+            ResolvedAtomTerm::Var(var) => {
+                if let Some((i, _ty)) = self.locals.get_full(var) {
+                    self.instructions.push(Instruction::Load(Load::Stack(i)));
+                } else {
+                    let (i, _, _ty) = self.types.get_full(&var.name).unwrap();
+                    self.instructions.push(Instruction::Load(Load::Subst(i)));
+                }
+            }
+            ResolvedAtomTerm::Literal(lit) => {
+                self.instructions.push(Instruction::Literal(lit.clone()));
+            }
+            ResolvedAtomTerm::Global(var) => {
+                assert!(self.egraph().global_bindings.contains_key(&var.name));
+                self.instructions.push(Instruction::Global(var.name));
+            }
+        }
+    }
+
+    fn do_function(&mut self, func_type: &FuncType) {
+        self.instructions.push(Instruction::CallFunction(
+            func_type.name,
+            func_type.has_default || func_type.is_datatype,
+        ));
+    }
+
+    fn do_prim(&mut self, prim: &SpecializedPrimitive) {
+        self.instructions.push(Instruction::CallPrimitive(
+            prim.primitive.clone(),
+            prim.input.len(),
+        ));
+    }
+}
+
+#[derive(Clone, Debug)]
+enum Load {
+    Stack(usize),
+    Subst(usize),
+}
+
+#[derive(Clone, Debug)]
+enum Instruction {
+    Literal(Literal),
+    Load(Load),
+    Global(Symbol),
+    // function to call, and whether to make defaults
+    CallFunction(Symbol, bool),
+    CallPrimitive(Primitive, usize),
+    DeleteRow(Symbol),
+    Set(Symbol),
+    Union(usize),
+    Extract(usize),
+    Panic(String),
+}
```

Comment:
I think documentation for each of the variants here would be helpful. Some of these _seem_ self-explanatory, but a short sentence explaining what each does would still be helpful. 

E.g. I believe these are all instructions manipulating a stack, but how they manipulate it is worth explaining here.